### PR TITLE
Support extended GraphRNN model

### PR DESCRIPTION
**Pull request by mike-shvets**
TODO:
- [x] BFSGraphDataset
- [x] GraphRNN training and inference modes
- [x] extended GraphRNN to num_node_classes > 2 and num_edge_classes > 2
- [x] fixed tensorboard logging crash on some machines
- [x] run on full chembl (needs cleanup: rare atoms, invalid molecules, etc.)
- [ ] one-hot embedding for unrelated classes
- [ ] SA score
- [ ] gradient barrier for hidden vector in inference mode
- [ ] handle class imbalance
- [ ] histograms for num_prev_nodes
- [ ] full probability decomposition options
- [ ] find best params for lr, network sizes, optimizer

### CODE REVIEW

File: configs/graphrnn_config.py [hunk 1]
```diff
@@ -9,6 +9,24 @@
...
 from openchem.modules.gru_plain import GRUPlain
+from openchem.utils.sa_score import sascorer
+from rdkit import Chem
+import numpy as np
+
+
+def get_sa_score(smiles, target=None):
+    scores = []
+    for sm in range(len(smiles)):
+        try:
```
Comment (mike-shvets) [inline configs/graphrnn_config.py]: which except cases do you expect here? Currently we do smiles cleanup in .forward(), so smiles should be valid. Does it work if we remove this?

File: configs/graphrnn_config.py [hunk 2]
```diff
@@ -9,6 +9,24 @@
...
+            mol = Chem.MolFromSmiles(sm)
+            if mol is not None and sm != '':
+                scores.append(sascorer.calculateScore(mol))
+            else:
+                scores.append(-1.0)
```
Comment (mike-shvets) [inline configs/graphrnn_config.py]: do we want to do this? Maybe only account for good ones?

File: configs/graphrnn_config.py [hunk 3]
```diff
@@ -93,19 +111,25 @@ def get_edge_attributes(bond):
...
-            'full_chembl_cleaned.pkl',
+    #pickled='benchmark_datasets/chembl_full/' +
+    #        'full_chembl_cleaned.pkl',
     get_bond_attributes=get_edge_attributes,
     edge_attributes=edge_attributes,
     delimiter=',',
     random_order=True, max_prev_nodes=max_prev_nodes,
     original_start_node_label=original_start_node_label,
     edge_relabel_map=edge_relabel_map,
+<<<<<<< Updated upstream
```
Comment (mike-shvets) [inline configs/graphrnn_config.py]: some artifacts from merge

File: openchem/data/utils.py [hunk 4]
```diff
@@ -12,6 +12,26 @@
...
+    def __iter__(self):
+        return self
+
+    def __len__(self):
+        return self.batch_size
+
+    def __next__(self):
+        if self.current < self.batch_size:
+            self.current += 1
+            return None
```
Comment (mike-shvets) [inline openchem/data/utils.py]: should this return a batch, smth like
batch = {
    "x": torch.zeros(...)
    "y": torch.zeros(...)
}
in order to match the output of BFSGraphDataset (i.e. model's cast_inputs don't break)

File: openchem/data/utils.py [hunk 5]
```diff
@@ -54,9 +74,13 @@ def pad_sequences(seqs, max_length=None, pad_symbol=' '):
...
                   pin_memory=False, sampler=None):
-    data_loader = DataLoader(dataset=dataset, batch_size=batch_size,
-                             shuffle=shuffle, num_workers=num_workers,
-                             pin_memory=pin_memory, sampler=sampler)
+    if dataset is not None:
+        data_loader = DataLoader(dataset=dataset, batch_size=batch_size,
+                                 shuffle=shuffle, num_workers=num_workers,
+                                 pin_memory=pin_memory, sampler=sampler)
+    else:
+        data_loader = DummyDataLoader(batch_size=batch_size)
```
Comment (mike-shvets) [inline openchem/data/utils.py]: if we create a dummy dataset directly in config so that it matches the BFSGraphDataset, we don't need to spoil code here

File: openchem/data/utils.py [hunk 6]
```diff
@@ -101,7 +125,7 @@ def sanitize_smiles(smiles, canonize=True,
                 idx.append(i)
                 num_atoms.append(n)
         except: 
-            warnings.warn('Unsanitized SMILES string: ' + sm)
+            #warnings.warn('Unsanitized SMILES string: ' + sm)
```
Comment (mike-shvets) [inline openchem/data/utils.py]: Can we just throw one warning at the end of the function if not all are valid (check len(idx) == len(smiles))

File: openchem/data/utils.py [hunk 7]
```diff
@@ -132,8 +156,8 @@ def canonize_smiles(smiles, sanitize=True):
                 Chem.MolToSmiles(Chem.MolFromSmiles(sm, sanitize=sanitize))
             )
         except: 
-            warnings.warn(sm + ' can not be canonized: i'
-                                'nvalid SMILES string!')
+            #warnings.warn(sm + ' can not be canonized: i'
```
Comment (mike-shvets) [inline openchem/data/utils.py]: same here

File: openchem/models/openchem_model.py [hunk 8]
```diff
@@ -233,15 +234,28 @@ def evaluate(model, val_loader, criterion):
         eval_metrics = model.eval_metrics
         world_size = model.world_size
     for i_batch, sample_batched in enumerate(val_loader):
-        if has_module:
-            batch_input, batch_target = model.module.cast_inputs(sample_batched)
+        if sample_batched is not None:
```
Comment (mike-shvets) [inline openchem/models/openchem_model.py]: again, if dummy dataset is nice, we don't have to modify here

File: run.py [hunk 9]
```diff
@@ -207,12 +207,12 @@ def main():
     else:
         train_loader = None
 
-    if args.mode in ["eval", "train_eval"] and (
-            'val_data_layer' not in model_config.keys()
-            or model_config['val_data_layer'] is None):
-        raise IOError(
-            "When model is run in 'eval' or 'train_eval' modes, "
-            "validation data layer must be specified")
+    #if args.mode in ["eval", "train_eval"] and (
```
Comment (mike-shvets) [inline run.py]: why is this gone?

File: openchem/data/utils.py (hunk 4 above)
Comment (mike-shvets) [reply openchem/data/utils.py]: Actually, I thought this is a dataset first. Maybe leave the default data loader and only create a dataset that matches the signature of BFSGraphDataset, but with all zero tensors? Those are not used in model inference anyway

File: openchem/models/openchem_model.py [hunk 10]
```diff
@@ -233,15 +234,28 @@ def evaluate(model, val_loader, criterion):
...
+            if has_module:
+                batch_input, batch_target = model.module.cast_inputs(sample_batched)
+            else:
+                batch_input, batch_target = model.cast_inputs(sample_batched)
         else:
-            batch_input, batch_target = model.cast_inputs(sample_batched)
+            batch_input = None
+            batch_target = [None]
+        model.eval()
```
Comment (mike-shvets) [inline openchem/models/openchem_model.py]: Can move model.eval() to the top

File: openchem/models/openchem_model.py [hunk 11]
```diff
@@ -233,15 +234,28 @@ def evaluate(model, val_loader, criterion):
...
         predicted = model.forward(batch_input, eval=True)
-        prediction += list(predicted.detach().cpu().numpy())
-        ground_truth += list(batch_target.cpu().numpy())
+        if hasattr(predicted, 'detach'):
+            predicted = predicted.detach().cpu().numpy()
+        if hasattr(batch_target, 'cpu'):
+            batch_target = batch_target.cpu().numpy()
+        prediction += list(predicted)
+        ground_truth += list(batch_target)
         loss = criterion(predicted, batch_target)
```
Comment (mike-shvets) [inline openchem/models/openchem_model.py]: actually we can also redefine DummyLoss to return torch.zeros(0, dtype=torch.float) if the input is not torch.Tensor
